File: src/main/java/org/example/kalkulationsprogramm/controller/EmailTemplateController.java
package org.example.kalkulationsprogramm.controller;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

import org.example.email.EmailService;
import org.example.kalkulationsprogramm.service.DokumentFreigabeService;
import org.example.kalkulationsprogramm.service.EmailTextTemplateService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import lombok.Data;
import lombok.RequiredArgsConstructor;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/email/template")
public class EmailTemplateController {

    private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd");
    private static final DateTimeFormatter DISPLAY_DATE_FORMAT = DateTimeFormatter.ofPattern("dd.MM.yyyy");
    private static final String REVIEW_LINK = "<a href='https://www.google.com/search?q=Bauschlosserei+Thomas+Kuhn+Rezensionen#lkt=LocalPoiReviews'>Jetzt Bewertung abgeben</a>";

    private final EmailTextTemplateService emailTextTemplateService;
    private final DokumentFreigabeService dokumentFreigabeService;

    @PostMapping
    public ResponseEntity<EmailTemplateResponse> generateTemplate(@RequestBody EmailTemplateRequest request) {
        if (request.getDokumentTyp() == null || request.getDokumentTyp().isBlank()) {
            return ResponseEntity.badRequest().build();
        }

        String dokumentTyp = request.getDokumentTyp().toUpperCase();

        try {
            EmailService.EmailContent content = renderFromDbTemplate(dokumentTyp, request);
            if (content == null) {
                content = renderFallback(dokumentTyp, request);
            }

            String body = content.htmlBody();

            // Nur Angebote bekommen einen digitalen Freigabe-Link.
            // Auftragsbestätigungen werden vom Büro versendet — der Kunde hat dem Angebot
            // bereits zugestimmt, eine zweite digitale Bestätigung waere irreführend.
            if (dokumentTyp.equals("ANGEBOT")
                    && request.getDokumentId() != null) {
                boolean isAnfrage = Boolean.TRUE.equals(request.getIsAnfrage());
                String recipient = request.getRecipient() != null ? request.getRecipient() : "";
                String pdfDateiname = request.getPdfDateiname();
                int gueltigkeitTage = request.getGueltigkeitTage() != null
                        ? request.getGueltigkeitTage()
                        : DokumentFreigabeService.DEFAULT_GUELTIGKEITS_TAGE;
                body += dokumentFreigabeService
                        .erstelleFreigabeBlockFuerDokument(request.getDokumentId(), isAnfrage, recipient, pdfDateiname, gueltigkeitTage)
                        .orElse("");
            }

            EmailTemplateResponse response = new EmailTemplateResponse();
            response.setSubject(content.subject());
            response.setBody(body);
            return ResponseEntity.ok(response);

        } catch (Exception e) {
            return ResponseEntity.internalServerError().build();
        }
    }

    private EmailService.EmailContent renderFromDbTemplate(String dokumentTyp, EmailTemplateRequest req) {
        Map<String, String> ctx = new HashMap<>();
        ctx.put("ANREDE", nullToEmpty(req.getAnrede(), "Sehr geehrte Damen und Herren"));
        ctx.put("KUNDENNAME", nullToEmpty(req.getKundenName(), ""));
        ctx.put("ANSPRECHPARTNER", nullToEmpty(req.getAnsprechpartner(), ""));
        ctx.put("BAUVORHABEN", nullToEmpty(req.getBauvorhaben(), ""));
        ctx.put("PROJEKTNUMMER", nullToEmpty(req.getProjektnummer(), ""));
        ctx.put("DOKUMENTNUMMER", nullToEmpty(req.getDokumentnummer(), ""));
        ctx.put("BENUTZER", nullToEmpty(req.getBenutzer(), ""));
        ctx.put("BETRAG", nullToEmpty(req.getBetrag(), ""));
        ctx.put("RECHNUNGSDATUM", formatDate(req.getRechnungsdatum()));
        ctx.put("FAELLIGKEITSDATUM", formatDate(req.getFaelligkeitsdatum()));
        ctx.put("REVIEW_LINK", REVIEW_LINK);

        return emailTextTemplateService.render(dokumentTyp, ctx);
    }

    private EmailService.EmailContent renderFallback(String dokumentTyp, EmailTemplateRequest request) {
        String anrede = request.getAnrede() != null ? request.getAnrede() : "Sehr geehrte Damen und Herren";
        String bauvorhaben = request.getBauvorhaben() != null ? request.getBauvorhaben() : "";
        String projektnummer = request.getProjektnummer() != null ? request.getProjektnummer() : "";
        String benutzer = request.getBenutzer() != null ? request.getBenutzer() : "";

        switch (dokumentTyp) {
            case "RECHNUNG", "TEILRECHNUNG", "SCHLUSSRECHNUNG", "ABSCHLAGSRECHNUNG" -> {
                String dokumentnummer = request.getDokumentnummer() != null ? request.getDokumentnummer() : "";
                LocalDate rechnungsdatum = parseDate(request.getRechnungsdatum(), LocalDate.now());
                LocalDate faelligkeitsdatum = parseDate(request.getFaelligkeitsdatum(), LocalDate.now().plusDays(14));
                String betrag = request.getBetrag() != null ? request.getBetrag() : "0,00 €";
                String kundenName = request.getKundenName() != null ? request.getKundenName() : "";
                return EmailService.buildInvoiceEmailWithTypeHints(
                        dokumentTyp.toLowerCase(), anrede, kundenName, bauvorhaben, projektnummer,
                        dokumentnummer, rechnungsdatum, faelligkeitsdatum, betrag, benutzer, dokumentTyp);
            }
            case "MAHNUNG" -> {
                String dokumentnummer = request.getDokumentnummer() != null ? request.getDokumentnummer() : "";
                LocalDate rechnungsdatum = parseDate(request.getRechnungsdatum(), LocalDate.now());
                LocalDate faelligkeitsdatum = parseDate(request.getFaelligkeitsdatum(), LocalDate.now().plusDays(14));
                String betrag = request.getBetrag() != null ? request.getBetrag() : "0,00 €";
                String kundenName = request.getKundenName() != null ? request.getKundenName() : "";
                return EmailService.buildInvoiceEmailWithTypeHints(
                        "mahnung", anrede, kundenName, bauvorhaben, projektnummer, dokumentnummer,
                        rechnungsdatum, faelligkeitsdatum, betrag, benutzer, "MAHNUNG");
            }
            case "ANGEBOT" -> {
                String anfragesnummer = request.getDokumentnummer() != null ? request.getDokumentnummer() : "";
                String kundenName = request.getKundenName() != null ? request.getKundenName() : "";
                return EmailService.buildOfferEmail(anrede, kundenName, bauvorhaben, anfragesnummer, benutzer, null);
            }
            case "AUFTRAGSBESTAETIGUNG" -> {
                String auftragsnummer = request.getDokumentnummer() != null ? request.getDokumentnummer() : projektnummer;
                String betrag = request.getBetrag() != null ? request.getBetrag() : null;
                String kundenName = request.getKundenName() != null ? request.getKundenName() : "";
                return EmailService.buildOrderConfirmationEmail(
                        null, anrede, kundenName, bauvorhaben, projektnummer, auftragsnummer, betrag, benutzer);
            }
            case "ZEICHNUNG" -> {
                return EmailService.buildDrawingEmail(anrede, benutzer, bauvorhaben);
            }
            default -> {
                return new EmailService.EmailContent("", "");
            }
        }
    }

    private static String formatDate(String dateStr) {
        if (dateStr == null || dateStr.isBlank()) {
            return "";
        }
        try {
            return LocalDate.parse(dateStr, DATE_FORMAT).format(DISPLAY_DATE_FORMAT);
        } catch (Exception e) {
            return dateStr;
        }
    }

    private static String nullToEmpty(String value, String fallback) {
        if (value == null || value.isBlank()) {
            return fallback == null ? "" : fallback;
        }
        return value;
    }

    private LocalDate parseDate(String dateStr, LocalDate defaultValue) {
        if (dateStr == null || dateStr.isBlank()) {
            return defaultValue;
        }
        try {
            return LocalDate.parse(dateStr, DATE_FORMAT);
        } catch (Exception e) {
            return defaultValue;
        }
    }

    @Data
    public static class EmailTemplateRequest {
        private String dokumentTyp;
        private String anrede;
        private String kundenName;
        private String ansprechpartner;
        private String bauvorhaben;
        private String projektnummer;
        private String dokumentnummer;
        private String rechnungsdatum;
        private String faelligkeitsdatum;
        private String betrag;
        private String benutzer;
        /** Optionale Dokument-ID für Freigabe-Block (nur bei Angebot/AB) */
        private Long dokumentId;
        /** true = Anfrage-Kontext, false = Projekt-Kontext */
        private Boolean isAnfrage;
        /** Empfänger-E-Mail für die Freigabe-Zuordnung */
        private String recipient;
        /** Dateiname der gespeicherten PDF (UUID.pdf) für die Freigabe-Seite */
        private String pdfDateiname;
        /** Vom Anwender gewählte Gültigkeitsdauer des Freigabe-Links in Tagen (nur bei Angebot). */
        private Integer gueltigkeitTage;
    }

    @Data
    public static class EmailTemplateResponse {
        private String subject;
        private String body;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/EmailTemplateController.kt
package org.example.kalkulationsprogramm.controller

import org.example.email.EmailService
import org.example.kalkulationsprogramm.service.DokumentFreigabeService
import org.example.kalkulationsprogramm.service.EmailTextTemplateService
import org.example.kalkulationsprogramm.service.FirmeninformationService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

@RestController
@RequestMapping("/api/email/template")
class EmailTemplateController(
    private val emailTextTemplateService: EmailTextTemplateService,
    private val dokumentFreigabeService: DokumentFreigabeService,
    private val firmeninformationService: FirmeninformationService,
) {
    @PostMapping
    fun generateTemplate(@RequestBody request: EmailTemplateRequest): ResponseEntity<EmailTemplateResponse> {
        val requestedDokumentTyp = request.dokumentTyp
        if (requestedDokumentTyp.isNullOrBlank()) {
            return ResponseEntity.badRequest().build()
        }

        val dokumentTyp = requestedDokumentTyp.uppercase(Locale.ROOT)
        return try {
            val content = renderFromDbTemplate(dokumentTyp, request) ?: renderFallback(dokumentTyp, request)
            var body = content.htmlBody()

            if ((dokumentTyp == "ANGEBOT" || dokumentTyp == "NACHTRAGSANGEBOT") && request.dokumentId != null) {
                val isAnfrage = request.isAnfrage == true
                val recipient = request.recipient ?: ""
                val gueltigkeitTage = request.gueltigkeitTage ?: DokumentFreigabeService.DEFAULT_GUELTIGKEITS_TAGE
                body += dokumentFreigabeService
                    .erstelleFreigabeBlockFuerDokument(
                        request.dokumentId,
                        isAnfrage,
                        recipient,
                        request.pdfDateiname,
                        gueltigkeitTage,
                    )
                    .orElse("")
            }

            ResponseEntity.ok(EmailTemplateResponse(content.subject(), body))
        } catch (e: Exception) {
            ResponseEntity.internalServerError().build()
        }
    }

    private fun renderFromDbTemplate(dokumentTyp: String, req: EmailTemplateRequest): EmailService.EmailContent? {
        val ctx = mutableMapOf<String, String>()
        ctx["ANREDE"] = nullToEmpty(req.anrede, "Sehr geehrte Damen und Herren")
        ctx["KUNDENNAME"] = nullToEmpty(req.kundenName, "")
        ctx["ANSPRECHPARTNER"] = nullToEmpty(req.ansprechpartner, "")
        ctx["BAUVORHABEN"] = nullToEmpty(req.bauvorhaben, "")
        ctx["PROJEKTNUMMER"] = nullToEmpty(req.projektnummer, "")
        ctx["DOKUMENTNUMMER"] = nullToEmpty(req.dokumentnummer, "")
        ctx["BENUTZER"] = nullToEmpty(req.benutzer, "")
        ctx["BETRAG"] = nullToEmpty(req.betrag, "")
        ctx["RECHNUNGSDATUM"] = formatDate(req.rechnungsdatum)
        ctx["FAELLIGKEITSDATUM"] = formatDate(req.faelligkeitsdatum)
        ctx["REVIEW_LINK"] = buildReviewLink()
        return emailTextTemplateService.render(dokumentTyp, ctx)
    }

    private fun renderFallback(dokumentTyp: String, request: EmailTemplateRequest): EmailService.EmailContent {
        val anrede = request.anrede ?: "Sehr geehrte Damen und Herren"
        val bauvorhaben = request.bauvorhaben ?: ""
        val projektnummer = request.projektnummer ?: ""
        val benutzer = request.benutzer ?: ""

        return when (dokumentTyp) {
            "RECHNUNG", "TEILRECHNUNG", "SCHLUSSRECHNUNG", "ABSCHLAGSRECHNUNG" -> {
                val dokumentnummer = request.dokumentnummer ?: ""
                val rechnungsdatum = parseDate(request.rechnungsdatum, LocalDate.now())
                val faelligkeitsdatum = parseDate(request.faelligkeitsdatum, LocalDate.now().plusDays(14))
                val betrag = request.betrag ?: "0,00 €"
                val kundenName = request.kundenName ?: ""
                EmailService.buildInvoiceEmailWithTypeHints(
                    dokumentTyp.lowercase(Locale.ROOT),
                    anrede,
                    kundenName,
                    bauvorhaben,
                    projektnummer,
                    dokumentnummer,
                    rechnungsdatum,
                    faelligkeitsdatum,
                    betrag,
                    benutzer,
                    dokumentTyp,
                )
            }
            "MAHNUNG", "ZAHLUNGSERINNERUNG", "ERSTE_MAHNUNG", "ZWEITE_MAHNUNG" -> {
                val dokumentnummer = request.dokumentnummer ?: ""
                val rechnungsdatum = parseDate(request.rechnungsdatum, LocalDate.now())
                val faelligkeitsdatum = parseDate(request.faelligkeitsdatum, LocalDate.now().plusDays(14))
                val betrag = request.betrag ?: "0,00 €"
                val kundenName = request.kundenName ?: ""
                EmailService.buildInvoiceEmailWithTypeHints(
                    "mahnung",
                    anrede,
                    kundenName,
                    bauvorhaben,
                    projektnummer,
                    dokumentnummer,
                    rechnungsdatum,
                    faelligkeitsdatum,
                    betrag,
                    benutzer,
                    dokumentTyp,
                )
            }
            "ANGEBOT", "NACHTRAGSANGEBOT" -> {
                val anfragesnummer = request.dokumentnummer ?: ""
                val kundenName = request.kundenName ?: ""
                EmailService.buildOfferEmail(anrede, kundenName, bauvorhaben, anfragesnummer, benutzer, null)
            }
            "AUFTRAGSBESTAETIGUNG" -> {
                val auftragsnummer = request.dokumentnummer ?: projektnummer
                val kundenName = request.kundenName ?: ""
                EmailService.buildOrderConfirmationEmail(
                    null,
                    anrede,
                    kundenName,
                    bauvorhaben,
                    projektnummer,
                    auftragsnummer,
                    request.betrag,
                    benutzer,
                )
            }
            "ZEICHNUNG" -> EmailService.buildDrawingEmail(anrede, benutzer, bauvorhaben)
            else -> EmailService.EmailContent("", "")
        }
    }

    private fun buildReviewLink(): String {
        val url = firmeninformationService.firmeninformation?.googleBewertungsLink
        if (url.isNullOrBlank()) {
            return ""
        }
        val safeUrl = url.trim().replace("\"", "%22")
        return "<a href=\"$safeUrl\" target=\"_blank\" rel=\"noopener noreferrer\">$REVIEW_LINK_LABEL</a>"
    }

    private fun parseDate(dateStr: String?, defaultValue: LocalDate): LocalDate {
        if (dateStr.isNullOrBlank()) {
            return defaultValue
        }
        return try {
            LocalDate.parse(dateStr, DATE_FORMAT)
        } catch (e: Exception) {
            defaultValue
        }
    }

    data class EmailTemplateRequest(
        var dokumentTyp: String? = null,
        var anrede: String? = null,
        var kundenName: String? = null,
        var ansprechpartner: String? = null,
        var bauvorhaben: String? = null,
        var projektnummer: String? = null,
        var dokumentnummer: String? = null,
        var rechnungsdatum: String? = null,
        var faelligkeitsdatum: String? = null,
        var betrag: String? = null,
        var benutzer: String? = null,
        var dokumentId: Long? = null,
        var isAnfrage: Boolean? = null,
        var recipient: String? = null,
        var pdfDateiname: String? = null,
        var gueltigkeitTage: Int? = null,
    )

    data class EmailTemplateResponse(
        var subject: String? = null,
        var body: String? = null,
    )

    companion object {
        private val DATE_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
        private val DISPLAY_DATE_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("dd.MM.yyyy")
        private const val REVIEW_LINK_LABEL = "Jetzt Bewertung abgeben"

        private fun formatDate(dateStr: String?): String {
            if (dateStr.isNullOrBlank()) {
                return ""
            }
            return try {
                LocalDate.parse(dateStr, DATE_FORMAT).format(DISPLAY_DATE_FORMAT)
            } catch (e: Exception) {
                dateStr
            }
        }

        private fun nullToEmpty(value: String?, fallback: String?): String {
            return if (value.isNullOrBlank()) fallback ?: "" else value
        }
    }
}
